### ISSUE TITLE
Safe to disable X-FRAME-OPTIONS for Content Security Policy?

### ISSUE THREAD

**Issue by Sawtaytoes** [helmetjs/helmet#167]
Is it safe to disable `frameguard` and use `contentSecurityPolicy` instead? For some reason, Chrome is prioritizing `X-FRAME-OPTIONS`, and it doesn't have a way to allow an entire domain, only a specific URL.

This is the solution I came up with:
```js
helmet({
  contentSecurityPolicy: {
    directives: {
      defaultSrc: ["'self'"],
    },
  },
  frameguard: false,
})
```

Since I have an iframe I'm using for authentication on a variety of same-domain (but not same-subdomain) sites, I'm looking for a way of replacing `frameguard` with `contentSecurityPolicy` so I can set `frameSrc: ["'self'", '*.example.com']`

**Comment by EvanHahn** [comment]
I'm not exactly sure what your use case is here. `X-Frame-Options` is useful to prevent your page from being put _inside_ an iframe where `Content-Security-Policy` is useful to prevent errant iframes from being put on your page.

You could set [the `frame-ancestors` CSP directive](https://www.w3.org/TR/CSP2/#directive-frame-ancestors) in the iframe and the `frame-src` directive on the outer page. Would that work?

You should also be aware of browser support for both [`X-Frame-Options`](https://caniuse.com/#feat=x-frame-options) and [`Content-Security-Policy`](https://caniuse.com/#feat=contentsecuritypolicy).

**Comment by ryanfitzer** [comment]
@EvanHahn 

> I'm not exactly sure what your use case is here. X-Frame-Options is useful to prevent your page from being put inside an iframe where Content-Security-Policy is useful to prevent errant iframes from being put on your page.

From what I understand, your explanation is incorrect. The [frame-ancestors](https://www.w3.org/TR/CSP2/#frame-ancestors-and-frame-options) docs say:

> This directive is similar to the `X-Frame-Options` header that several user agents have implemented. The 'none' source expression is roughly equivalent to that header’s `DENY`, `'self'` to `SAMEORIGIN`, and so on. The major difference is that many user agents implement `SAMEORIGIN` such that it only matches against the top-level document’s location. This directive checks each ancestor. If any ancestor doesn’t match, the load is cancelled.

**Comment by EvanHahn** [comment]
Yes, I think you're correct. CSP's `frame-ancestors` is an improved version of the `X-Frame-Options` header. I think its main disadvantage would be legacy browser support, but I think that's probably minor.

Does that help?

**Comment by EvanHahn** [comment]
I'm going to close this issue due to inactivity, but please let me know if you have any more questions and I can reopen it.

**Comment by Sawtaytoes** [comment]
I don't understand how to answer your questions. Is there a better configuration I could've used which would've kept a higher security level?

**Comment by EvanHahn** [comment]
I think I don't quite understand your use case. Do you control both the outer page and the authentication iframe?

My understanding is this:
- You own the outer pages, which are all on different subdomains. For example, `foo.example.com`, `boo.example.com`, etc.
- You also own the authentication iframes, which are all on the same domain. For example, `auth-example-iframe.com`.

Is that understanding correct? If so, I can better help advise.

**Comment by Sawtaytoes** [comment]
I control both yes. Your understanding is correct! :D

**Comment by EvanHahn** [comment]
Okay—thanks!

**tl;dr, it's safe to disable `X-Frame-Options` as long as you also specify [the `frame-ancestors` Content Security Policy directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors), but be wary of the latter's browser support.**

At the end of the day, you have two goals:

1. Your outer pages should only allow `auth-example-iframe.com` to be put in an iframe.
2. Your iframe should only be embedded by `foo.example.com` and `boo.example.com`.

(Obviously these are example domains, and I'm only assuming you have two outer domains, but this should be easy to map to your actual setup.)

These two goals are largely unrelated in terms of implementation—you'll need to set two different policies in order to achieve this.

# Goal 1: only allow the auth iframe

Your outer pages will want a Content Security Policy that only allows `auth-example-iframe.com` to be framed. You can achieve this with the `frame-src` directive.

Here's what that might look like with Helmet:

```javascript
app.use(helmet.contentSecurityPolicy({
  directives: {
    // ...
    frameSrc: ['https://*.auth-example-iframe.com']
  }
}))
```

The more specific you can make that URL, the tighter your policy will be (which is generally a good thing). For example, if you know the exact URL, you can specify that:

```javascript
frameSrc: ['https://auth.auth-example-iframe.com/auth']
```

The `X-Frame-Options` header (`helmet.frameguard`) is irrelevant here. I'd recommend that you include it so that your outer pages don't get put in iframes, but that's a separate issue.

# Goal 2: only allow the iframe to be in some places

The iframe should only be embedded on `foo.example.com` and `boo.example.com`, let's say. You can achieve this with the `frame-ancestors` directive.

Because you want multiple different domains, the `X-Frame-Options` header should be removed, as you've done.

Here's what that might look like in code:

```javascript
app.use(helmet({
  contentSecurityPolicy: {
    directives: {
      frameAncestors: [
        'https://foo.example.com',
        'https://boo.example.com'
      ]
    }
  },
  frameguard: false
}))
```

If you wish to allow *any* subdomain of example.com:

```javascript
frameAncestors: ['https://*.auth-example-iframe.com']
```

Browser support might be a concern here. Content Security Policy Level 2 deprecates `frame-src` and introduces `child-src` and `frame-ancestors`. [Not all browsers support CSP 2](https://www.caniuse.com/#feat=contentsecuritypolicy2).

Does that help?

**Comment by Malvoz** [comment]
@EvanHahn 

Althought CSP2 is recommendation, I'm certain we should be looking at CSP3 at this point. I'm not sure why CSP2 is still recommendation. 

&lt;researches for a minute&gt; 
Oh! confirmed: https://github.com/w3c/webappsec-csp/issues/244#issuecomment-333099819

**Comment by EvanHahn** [comment]
I agree, but `X-Frame-Options` doesn't support this exact use case. You can hack it in various ways if IE support is important to you.

**Comment by coffy** [comment]
app.use(helmet({frameguard: false})) executes framequard middleware with option SAMEORIGIN. 
There is no way currently to configure helmet to avoid X-Frame-Options header.
Am I missing something?

**Comment by NathanCH** [comment]
To expand on @Malvoz's point, it's important to keep X-Frame-Options otherwise you're susceptible to attacks from legacy browsers as recent as IE9. It's recommended to use both X-Frame-Options and a CSP. X-Frame-Options will ignored by modern browsers in favor of a CSP.

...with one exception: Safari 12 still prioritizes X-Frame-Options. This is problematic if your CSP header whitelists multiple domains.

**Comment by coffy** [comment]
@EvanHahn I am not sure why but during my previous testing I did see frameguard to execute and to include header into response, but now I can't reproduce it. 
Thanks, it looks like you are right.

**Comment by EvanHahn** [comment]
It seems like this question has been largely answered, so I'm going to close this issue.

Feel free to continue discussion or ask me to reopen if needed!

**Comment by sahil290791** [comment]
@NathanCH Let's say I have **X-Frame-Options** as SAMEORIGIN and CSP's **frame-ancestors** directive has multiple domains whitelisted as my current requirement to load domain xyz in domain abc. How can I solve this problem?

**Comment by EvanHahn** [comment]
@sahil290791 According to [the spec](https://w3c.github.io/webappsec-csp/2/#frame-ancestors-and-frame-options), `frame-ancestors` takes priority.

> The `frame-ancestors` directive obsoletes the `X-Frame-Options` header. If a resource has both policies, the `frame-ancestors` policy SHOULD be enforced and the `X-Frame-Options` policy SHOULD be ignored.

**Comment by wootwoot1234** [comment]
@EvanHahn Thanks for the example code, it's very helpful.  I'm trying to follow your example and this is what I came up with.  I had to add in the default directives because I think not specifying them removed them.  I want everything to be default except for allowing an iFrame to be linked from a subdomain.  I'm wanting to be able to include the `https://example.com/register` page on a blog at `https://blog.example.com`.   Does it seem right to you?  

    app.use(
        helmet({
            contentSecurityPolicy: {
                directives: {
                    frameAncestors: ["https://blog.example.com"],
                    frameSrc: ["https://example.com/register"],
                    defaultSrc: "self",
                    baseUri: "self",
                    blockAllMixedContent: [],
                    objectSrc: "none",
                    scriptSrc: "self",
                    scriptSrcAttr: "none",
                    upgradeInsecureRequests: []
                }
            },
            frameguard: false
        })
    );

**Comment by EvanHahn** [comment]
@wootwoot1234 It seems like you want something like this:

- Loading `https://blog.example.com` should allow frames from `https://example.com/register`. In other words, it should set `frame-src https://example.com/register`.
- Loading `https://example.com/register` in a frame should be allowed if the outer frame is `https://blog.example.com`. In other words, it should set `frame-ancestors https://blog.example.com`.

It looks like your usage of Content Security Policy is grouped such that this will work, but other things are possible. For example, you'll allow `https://blog.example.com` to be framed by `https://blog.example.com`, and you'll allow `https://example.com/register` to load `https://example.com/register` in a frame. You'll also allow any other URLs with this header to do the same—so you might be allowing `https://example.com/foo-bar` to load `https://example.com/register` in a frame.

If that's what you want, great! If not, you might want to try using two different invocations of the middleware.

For example, something like this:

```js
routerThatHandlesExampleDotComSlashRegister.use(helmet({
  contentSecurityPolicy: {
    directives: {
      // ...
      frameAncestors: ["https://blog.example.com"],
    },
  },
  // ...
}));

routerThatHandlesBlogExampleDotCom.use(helmet({
  contentSecurityPolicy: {
    directives: {
      // ...
      frameSrc: ["https://example.com/register"],
    },
  },
  // ...
}));
```

That will set two different Content Security Policies, which I *think* is what you want.

Does that help?

**Comment by wootwoot1234** [comment]
Thank you @EvanHahn.  That makes a lot more sense.  Thanks!

**Comment by amarbalu** [comment]
@EvanHahn Beautiful explanation. So to avoid my web application being I-framed to prevent click jacking. X-Frame-Options is already in place. Do we need to still add frame-ancestors?

**Comment by EvanHahn** [comment]
> Do we need to still add frame-ancestors : self in our web-application? Is this necessary!!!

@amarbalu Yes, I would add both. Hope this helps!

**Comment by amarbalu** [comment]
@EvanHahn Thanks for your quick response. As per our discussion. Do we need to disable frame guard ( X-FRAME-OPTIONS) for adding frame-ancestors!!!

**Comment by EvanHahn** [comment]
Nope! It should be good to enable both.

**Comment by amarbalu** [comment]
Thanks @EvanHahn for your support.

**Comment by EvanHahn** [comment]
@yellowjacketcoder Yes, you should disable the `X-Frame-Options` header, because you want your page to be put into an iframe on another domain. You're right to use CSP here.

**Comment by yellowjacketcoder** [comment]
Thank you very much @EvanHahn
